src/mir/construction.md

```diff
@@ -1 +1,111 @@
 # MIR construction
+
+The lowering of [HIR] to [MIR] occurs for the following (probably incomplete)
+list of items:
+
+* Function and Closure bodies
+* Initializers of `static` and `const` items
+* Initializers of enum discriminants
+* Glue and Shims of any kind
+    * Tuple struct initializer functions
+    * Drop code (the `Drop::drop` function is not called directly)
+    * Drop implementations of types without an explicit `Drop` implementation
+
+The lowering is triggered by calling the `mir_built` query. The entire lowering
+code lives in `src/librustc_mir/build`. There is an intermediate representation
+between [HIR] and [MIR] called the `HAIR` that is only used during the lowering.
+The `HAIR` has datatypes that mirror the [HIR] datatypes, but instead of e.g. `-x`
+being a `hair::ExprKind::Neg(hair::Expr)` it is a `hair::ExprKind::Neg(hir::Expr)`.
+
+This shallowness enables the `HAIR` to represent all datatypes that [HIR] has, but
+without having to create an in-memory copy of the entire [HIR]. The `HAIR` also
+does a few simplifications, e.g. method calls and function calls have been merged
+into a single variant.
```

Comment:
The most important thing about the HAIR is that all "adjustments" and other implicit bits of syntax are made explicit, I think. Can we mention that?

Reply:
It might be worth linking to this module too <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/hair/cx/expr/index.html>, as that is where HAIR construction takes place. Kind of annoying: it's hard to get a good stable-ish URL to a trait impl.

Reply:
Another example: HAIR converts things like overloaded operators into fn calls